Add Public APIs for Media Renegotiation

jpsantosbh opened the pull request:
# Description

Add Public APIs for Media Renegotiation, allowing a developer to change the media options and execute a renegotiation.

## Type of change

- [ ] Internal refactoring
- [ ] Bug fix (bugfix - non-breaking)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Code snippets

```javascript
await call.renegotiateMedia({video: false, audio: true, negotiateVideo: true,  negotiateAudio: true,});
// or
await call.enableVideo({sendOnly: true});
// or
await call.disableVideo({recvOnly: true});
```

In case of new feature or breaking changes, please include code snippets.

iAmmar7 commented on packages/webrtc/src/RTCPeer.ts:
```diff
@@ -620,6 +623,10 @@ export default class RTCPeer<EventTypes extends EventEmitter.ValidEventTypes> {
   }
+  _disposeLocalStream() {
+    this._localStream?.getTracks().forEach((track) => track.stop())
+    this._localStream = undefined
```
Why set it to undefined? Also, you might want to use the `stopTrack` method to stop the audio track.

jpsantosbh replied:
This method releases the previously allocated resources, not only to stop the tracks.

iAmmar7 commented on packages/webrtc/src/BaseConnection.ts:
```diff
@@ -1086,4 +1081,18 @@ export class BaseConnection<EventTypes extends EventEmitter.ValidEventTypes>
+  async renegotiateMedia(renegotiateMediaParams: UpdateMediaOptions): Promise<void> {
+    this.updateMediaOptions(renegotiateMediaParams)
+    //starts with offering = true
+    await this.peer?.start(true)
```
For more clarity, can we use the object param rather than just a boolean param? Something like:
```ts
start({ isOffer: boolean })
```

iAmmar7 commented on packages/webrtc/src/BaseConnection.ts:
```diff
@@ -1086,4 +1081,18 @@ export class BaseConnection<EventTypes extends EventEmitter.ValidEventTypes>
+    await this.renegotiateMedia({video: enableVideoParam?.video ?? true, negotiateVideo: !enableVideoParam?.sendOnly})
+  }
+
+  async disableVideo(disableVideoParam?: {recvOnly?: boolean}): Promise<void> {
```
If all three methods are public, you might want to expose these via CallFabricRoomSessionContract. Also, the method params need a dedicated type and expose via the SDK.

iAmmar7 commented on packages/webrtc/src/BaseConnection.ts:
```diff
@@ -1086,4 +1081,18 @@ export class BaseConnection<EventTypes extends EventEmitter.ValidEventTypes>
+    await this.peer?.start(true)
+  }
+
+  async enableVideo(enableVideoParam?: Pick<UpdateMediaOptions, 'video'> & {sendOnly?: boolean}): Promise<void> {
```
Can we set the param as `sendOnly` and `recvOnly` instead of a `video` param just to have more clarity?

jpsantosbh replied:
I'm not sure what you mean.
The `video` parameter is to specify a media constraint to be used by `gUM` while `sendOnly` is to control the negotiation.

iAmmar7 replied:
Makes sense, thanks! 👍

iAmmar7 commented:
Can you please integrate the API in the playground so that we can reproduce and see the implementation in action?

jpsantosbh commented on internal/e2e-js/tests/callfabric/renegotiation.spec.ts:
```diff
@@ -0,0 +1,226 @@
+    expect(stats.inboundRTP).toHaveProperty('video')
+
+    await page.waitForTimeout(1000)
+    expectMCUVisibleForAudience(page)
```
No, in this test, we do not expect the local overlay to be visible, but we expect the remote video to be visible.

iAmmar7 replied:
But why? The local overlay should be visible unless the user passes the applyLocalOverlay flag set to false, no?

jpsantosbh replied:
No, the local overlay is only created when you have a video track in the localStream. In this test case, we are `recvOnly`. There is no video track and no local overlay.

jpsantosbh commented:
I added the feature to the playground, also the Video SDK e2e are passing.

iAmmar7 commented:
> I added the feature to the playground, also the Video SDK e2e are passing.

Thanks for the playground update.

The Video SDK e2e would pass because it's a new feature. We might want to add a new e2e test for it as well, or we can decide to move functions to the CallFabricRoomSession class to ensure they are only exposed to the CF SDK for now.

iAmmar7 commented on packages/webrtc/src/BaseConnection.ts:
```diff
@@ -1086,4 +1081,17 @@ export class BaseConnection<EventTypes extends EventEmitter.ValidEventTypes>
     this.rtcPeerMap.clear()
   }
+
+  async renegotiateMedia(renegotiateMediaParams: UpdateMediaOptions): Promise<void> {
```
nit:
```suggestion
  async renegotiateMedia(params: UpdateMediaOptions): Promise<void> {
```

iAmmar7 commented on packages/webrtc/src/BaseConnection.ts:
```diff
@@ -1086,4 +1081,17 @@ export class BaseConnection<EventTypes extends EventEmitter.ValidEventTypes>
+    await this.peer?.start({isRenegotiate: true})
+  }
+
+  async enableVideo(enableVideoParam?: Pick<UpdateMediaOptions, 'video'> & {sendOnly?: boolean}): Promise<void> {
```
nit:
```suggestion
  async enableVideo(params?: Pick<UpdateMediaOptions, 'video'> & {sendOnly?: boolean}): Promise<void> {
```

Also, please declare the param type separately, so that it can be exposed to the public since it's a public method.

iAmmar7 commented on packages/webrtc/src/BaseConnection.ts:
```diff
@@ -1086,4 +1081,17 @@ export class BaseConnection<EventTypes extends EventEmitter.ValidEventTypes>
+    await this.renegotiateMedia({video: enableVideoParam?.video ?? true, negotiateVideo: !enableVideoParam?.sendOnly})
+  }
+
+  async disableVideo(disableVideoParam?: {recvOnly?: boolean}): Promise<void> {
```
nit: for consistency across public methods.

```suggestion
  async disableVideo(params?: {recvOnly?: boolean}): Promise<void> {
```

Please declare param types as well.

iAmmar7 commented:
A few observations I noticed while testing the feature in the playground;

While joining a room with two members I noticed that ;

1. If one member requests for `enableVideo` with the `sendOnly` flag set to `true`, it does not negotiate the video, it does not even open the user camera. I am not sure if this flag adds any value.
2. If one member enables the video and then later disables the video, I get this error and call disconnects
```json
{
    "code": "ICE_GATHERING_FAILED",
    "message": "Ice gathering timeout"
}
```

Lastly, which is more important, in my opinion, is the usage of the `start` function from the `RTCPeer`. Now that I can see its full usage, I am forced to think if this is even required. I thought there was a new server API that we needed to hit but it seems we only need to send the `verto.modify` which is already being integrated within the `BaseConnection` methods.

The method named `updateConstraints` is responsible for upgrading/downgrading the media stream, and updating those streams with a more robust recursive function which includes falling back to the previous constraints if the new constraints fail. It also handles the event listeners such as `camera.updated`, `microphone.updated`, etc, and calls the server API `verto.modify` indirectly.

You can clearly see the difference in your code if you simply comment down one line and add one new line;
```ts
async renegotiateMedia(params: UpdateMediaOptions): Promise<void> {
    this.updateMediaOptions(params)
    // await this.peer?.start({ isRenegotiate: true })
    await this.updateConstraints(params) // new line
}
```

jpsantosbh commented on internal/e2e-js/tests/callfabric/renegotiation.spec.ts:
```diff
@@ -0,0 +1,226 @@
+
+    expect(stats.outboundRTP).toHaveProperty('video')
+    expect(stats.inboundRTP).not.toHaveProperty('video')
+  })
```
In this case, the video element is not added at all. Is a send only

iAmmar7 replied:
Oh yeah, I see your other comment now, thanks!

Maybe we can assert the MCU element on the other member side since the enable video is with the sendOnly param set to true.

jpsantosbh commented:
> The Video SDK e2e would pass because it's a new feature. We might want to add a new e2e test for it, or we can move functions to the CallFabricRoomSession class to ensure they are only exposed to the CF SDK for now.

Yes, I meant it's not causing a regression in the Video SDK. 

The product requested the feature to be available on Both SDK.

jpsantosbh commented:
> A few observations I noticed while testing the feature in the playground;
> 
> While joining a room with two members I noticed that ;
> 
> 1. If one member requests for `enableVideo` with the `sendOnly` flag set to `true`, it does not negotiate the video, it does not even open the user camera. I am not sure if this flag adds any value.
> 2. If one member enables the video and then later disables the video, I get this error and call disconnects
> 
> ```json
> {
>     "code": "ICE_GATHERING_FAILED",
>     "message": "Ice gathering timeout"
> }
> ```
> 
> Lastly, which is more important, in my opinion, is the usage of the `start` function from the `RTCPeer`. Now that I can see its full usage, I am forced to think if this is even required. I thought there was a new server API that we needed to hit but it seems we only need to send the `verto.modify` which is already being integrated within the `BaseConnection` methods.
> 
> The method named `updateConstraints` is responsible for upgrading/downgrading the media stream, and updating those streams with a more robust recursive function which includes falling back to the previous constraints if the new constraints fail. It also handles the event listeners such as `camera.updated`, `microphone.updated`, etc, and calls the server API `verto.modify` indirectly.
> 
> You can clearly see the difference in your code if you simply comment down one line and add one new line;
> 
> ```ts
> async renegotiateMedia(params: UpdateMediaOptions): Promise<void> {
>     this.updateMediaOptions(params)
>     // await this.peer?.start({ isRenegotiate: true })
>     await this.updateConstraints(params) // new line
> }
> ```

I couldn't reproduce the issue you reported about `sendOnly` flag:
![Signalwire_Call_Demo](https://github.com/user-attachments/assets/bef65563-ce48-4fbb-b32c-4125e5a32775)
For reference, This is also covered by the 2e2 tests here: https://github.com/signalwire/signalwire-js/pull/1132/files#diff-e1d7afe85b7f7137519ee6170abc4c8cbbd5d3af7e2852666fd8b6f36f3ac103R110

I've noticed the ICE_GATHERING_FAILED issue, too, and I'm investigating, but it doesn't seem to be a client/SDK issue.

Regarding the suggestion of `this.updateConstraints(params)`, it doesn't produce the same results in many cases.
But the update events are a good point. I don't want to make changes to `updateConstraints`, but doing the events makes sense.

jpsantosbh commented:
I'm trying some changes to use the updateConstraints

jpsantosbh commented:
@iAmmar7 Tests are green with the changes required to use the `updateConstraints` instead of `start` (although, in some cases, I have to call `start` internally). They look safe, IMO. Let me know if you see any potential issues.

iAmmar7 commented on internal/e2e-js/tests/callfabric/renegotiation.spec.ts:
```diff
@@ -0,0 +1,185 @@
+    await createCFClient(page)
+
+    // Dial an address with audio only channel
+    const roomSession = await page.evaluate(
+      async ({ roomName }) => {
+        return new Promise<any>(async (resolve, _reject) => {
+          // @ts-expect-error
+          const client = window._client
+
+          const call = await client.dial({
+            to: `/public/${roomName}?channel=audio`,
+            rootElement: document.getElementById('rootElement'),
+          })
+
+          call.on('room.joined', resolve)
+          call.on('room.updated', () => {})
+
+          
+
+          // @ts-expect-error
+          window._roomObj = call
+
+          await call.start()
+        })
+      },
+      { roomName }
+    )
```
Forgot to update this to `dialAddress`?

jpsantosbh replied:
Why?

iAmmar7 replied:
We have a utility function `dialAddress` for this dialing. You are using it in the other two tests.

iAmmar7 commented on packages/js/src/utils/interfaces/callFabric.ts:
```diff
@@ -37,4 +38,25 @@ export interface CallFabricRoomSessionConnectionContract {
+   * 
+   * @param renegotiateMediaParams 
+   */
+  renegotiateMedia(renegotiateMediaParams: UpdateMediaOptions): Promise<void>
```
For consistency across the public methods exposed via `CallFabricRoomSession`, can you please rename these parameter variable name to `params`?

Public methods can be seen here: https://github.com/signalwire/signalwire-js/blob/fca4c09ac531ab88dec9d94f3a73d5cd06060d36/packages/js/src/fabric/CallFabricRoomSession.ts#L205

iAmmar7 commented on packages/js/src/utils/interfaces/callFabric.ts:
```diff
@@ -37,4 +38,25 @@ export interface CallFabricRoomSessionConnectionContract {
+   * 
+   * @param enableVideoParam 
+   */
+  enableVideo(enableVideoParam?: Pick<UpdateMediaOptions, 'video'> & {sendOnly?: boolean}): Promise<void>
```
Again for consistency across the methods.
```ts
  type EnableVideoParams = Pick<UpdateMediaOptions, 'video'> & { sendOnly?: boolean }
```

You may want to declare this type here since it's for both Video and CF SDK: https://github.com/signalwire/signalwire-js/blob/fca4c09ac531ab88dec9d94f3a73d5cd06060d36/packages/js/src/utils/interfaces/index.ts

And also please expose it from here for the Developer accessibility;
https://github.com/signalwire/signalwire-js/blob/fca4c09ac531ab88dec9d94f3a73d5cd06060d36/packages/js/src/index.ts

```suggestion  
  enableVideo(params?: EnableVideoParams): Promise<void>
```

iAmmar7 commented on packages/js/src/utils/interfaces/callFabric.ts:
```diff
@@ -37,4 +38,25 @@ export interface CallFabricRoomSessionConnectionContract {
+   * 
+   * @param enableVideoParam 
+   */
+  disableVideo(disableVideoParam?: {recvOnly?: boolean}): Promise<void>
```
Same for this method

iAmmar7 commented on internal/e2e-js/tests/callfabric/renegotiation.spec.ts:
```diff
@@ -0,0 +1,185 @@
+    await page.waitForTimeout(1000)
+
+    let stats = await getStats(page)
+0
```
Probably a mistake?
```suggestion
```

iAmmar7 commented on packages/webrtc/src/RTCPeer.ts:
```diff
@@ -421,7 +421,25 @@ export default class RTCPeer<EventTypes extends EventEmitter.ValidEventTypes> {
+          this.instance.getTransceivers().forEach((transceiver) => {
+            const kind = transceiver.receiver.track.kind
+            if(kind == 'audio' && this.options.audio && this.options.negotiateAudio) {
+              transceiver.direction = 'sendrecv'
```
Possible to set this direction in the `updateStream` method? Considering the `updateStream` is triggered by the `updateConstraints`, it is being used by our three new methods.

jpsantosbh replied:
The logic should be here.  This isn't a renegotiation-only concern.

This is the equivalent logic when the RTCPeerConnection supports transceivers.
Before, we were only handled when RTCPeerConnection didn't support transceivers.

The `startNegotiation`  is triggered by other cases that should also consider the transceivers in the negotiation.

iAmmar7 replied:
Gotcha! The `instance.addTrack` inside the updateStream method triggers the `negotiationneeded` event which calls the `startNegotiation`. So, it's already calling this method indirectly.

One or two if conditions can be merged into one, but I like your current approach, since it is more readable and cleaner. 👍

iAmmar7 reviewed (commented):
I have also shared a video regarding my findings for the `sendOnly` flag with you.

jpsantosbh commented on packages/webrtc/src/BaseConnection.ts:
```diff
@@ -515,6 +507,10 @@ export class BaseConnection<EventTypes extends EventEmitter.ValidEventTypes>
         }
 
         await this.updateStream(newStream)
+        if((!this.options.video && this.options.negotiateVideo) || (!this.options.audio && this.options.negotiateAudio)) {
```
Yes, This is a case where the existing RTCPeerConnection fires the needed negotiation event.
When we are "removing" one or both media channels we need create a new RTCPeerConnection.

iAmmar7 replied:
> When we are "removing" one or both media channels we need create a new RTCPeerConnection.

Umm.. .why? We can just update the current RTCPeer, can't we? As we are doing already in the case of attaching audio/video via the `updateConstraints` method.

For eg; if we join the call without video and do not negotiate it, and then later call the `updateConstraints` method with the video params, it will update the currently running RTC peer and send the `verto.modify` as well for renegotiation. Will that be helpful?

iAmmar7 replied:
The RTC peer automatically emits the [negotiationneeded](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/negotiationneeded_event) event, every time the track is being added via the `addTrack` method or removed via the `removeTrack` method.
https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/removeTrack

The observation you might have seen is because, the `updateStream` is calling the `addTrack` when you add the new track but it is not calling the `removeTrack` when you remove it. We might need to see where should we call this removeTrack method, we can't just call the `peer.start` method. It is not meant for renegotiation.

jpsantosbh replied:
>[ If the track is already stopped, or is not in the connection's senders list, this method has no effect.](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/removeTrack)

Won't work... I see no other way. Why are you so concerned about using `start()` here?

iAmmar7 replied:
There can be many reasons;
- Separation of concerns,
- Possibility of double renegotiation (remember that the `updateStream` triggers the renegotiation if require),
- Possibility of unnecessary renegotiation (if video changes but the negotiateVideo flag is still the same), it is possible that both the `updateStream` and the `updateConstraints` due to this `start` method begin the renegotiation,
- Unnecessarily restarting the peer connection may cause media interruptions plus resource usage.

I shared with you the 4 requirements of this PR, if you could please share with me which one of those requirements is failing, I would be able to debug and might be able to find a different solution.

jpsantosbh replied:
This is the test that will fail: https://github.com/signalwire/signalwire-js/pull/1132/files#diff-e1d7afe85b7f7137519ee6170abc4c8cbbd5d3af7e2852666fd8b6f36f3ac103R113

Another option requires more changes on the `startSegotiation`... Since the code already considers the double negotiation IMO, calling `start()` is cleaner. But I'll make the change in `startNegotiation`

jpsantosbh replied:
I tested changes to `startNegotiation`
 
But even with the correct SDPs, the test fails(we don't receive video)
The only way to do this is to create a new RTPPeerConnection.


The test scenario is that the user connects with `audio-only` but wants to start receiving video, too.
`enableVideo({video: false})` or `enableVideo({video: false, sendOnly:false})`   

https://github.com/signalwire/signalwire-js/pull/1132/files#diff-e1d7afe85b7f7137519ee6170abc4c8cbbd5d3af7e2852666fd8b6f36f3ac103R113

iAmmar7 replied:
This seems like a new requirement that does not fall under any of the above requirements mentioned. I think I tried to convey this before as well; before starting to implement the solution, we should first clearly write the whole problem. Also, the `enableVideo` with false video param does not make any sense.

In the meantime, I will try to see how can we solve this `recvOnly` problem.
